src/System.Memory/src/System/Span.cs

```diff
@@ -0,0 +1,413 @@
+        /// </summary>
+        /// <param name="obj">The managed object that contains the data to span over.</param>
+        /// <param name="rawPointer">A reference to data within that object.</param>
+        /// <param name="length">The number of <typeparamref name="T"/> elements the memory contains.</param>
+        /// <exception cref="System.ArgumentException">
+        /// Thrown when the specified object is an array or string. Use the type-specific constructor
+        /// for these special cases.
+        /// </exception>
+        /// <exception cref="System.ArgumentNullException">
+        /// Thrown when the specified object is null.
+        /// </exception>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="length"/> is negative.
+        /// </exception>
+        public static Span<T> DangerousCreate(object obj, ref T rawPointer, int length)
+        {
+            if (obj == null)
+                throw new ArgumentNullException(nameof(obj));
+            if (length < 0)
+                throw new ArgumentOutOfRangeException(nameof(length));
+
+            // Disallow "special" objects that have their own object headers and would yield unexpected (and runtime-dependent) results.
+            // We already have constructors for arrays (and that enforce the passing of the properly typed single-dimensional array.)
+            // This constructor is not useful for strings (cannot express"ref s[i]").
+            if (obj is Array)
+                throw new ArgumentException(SR.Argument_CannotPassArrayToDangerousCreate);
+            if (obj is string)
+                throw new ArgumentException(SR.Argument_CannotPassStringToDangerousCreate);
+
+            ref T pinnableReference = ref Unsafe.As<byte, T>(ref obj.GetPinnableReferenceForPlainObject());
+            IntPtr byteOffset = Unsafe.ByteOffset<T>(ref pinnableReference, ref rawPointer);
+            return new Span<T>(obj, byteOffset, length);
+        }
+
+        // Constructor for internal use only.
+        private Span(object o, IntPtr byteOffset, int length)
+        {
+            Debug.Assert(length >= 0);
+
+            Length = length;
+            _object = o;
+            _byteOffset = byteOffset;
+        }
+
+        /// <summary>
+        /// The number of items in the span.
+        /// </summary>
+        public int Length { get; }
+
+        /// <summary>
+        /// Returns true if Length is 0. 
+        /// </summary>
+        public bool IsEmpty => Length == 0;
+
+        /// <summary>
+        /// Returns a reference to specified element of the Span.
+        /// </summary>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        /// <exception cref="System.IndexOutOfRangeException">
+        /// Thrown when index less than 0 or index greater than or equal to Length
+        /// </exception>
+        public ref T this[int index]
```

Comment:
xproj and project.json. Everything else is using that and we'd need to make a new build with C# 7 which nobody is looking at. The next version of Visual Studio will have proper support.... but I'd like to use it with VS2015 and xproj since all of the existing .NET Core tooling is based on that.